File: Medium/2026-01-27-3650-Minimum-Cost-Path-with-Edge-Reversals/solution.java
import java.util.*;

class Solution {
    public int minimumCost(int n, int[][] edges) {
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int[] edge : edges) {
            graph.get(edge[0]).add(new int[]{edge[1], edge[2]});
        }
        
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[]{0, 0, 0}); // cost, node, used switch
        
        int[][] dist = new int[n][2];
        for (int[] d : dist) {
            Arrays.fill(d, Integer.MAX_VALUE);
        }
        dist[0][0] = 0;
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int cost = curr[0], node = curr[1], usedSwitch = curr[2];
            
            if (node == n - 1) {
                return cost;
            }
            
            if (cost > dist[node][usedSwitch]) {
                continue;
            }
            
            // Traverse normal edges
            for (int[] neighbor : graph.get(node)) {
                int nextNode = neighbor[0], edgeCost = neighbor[1];
                if (cost + edgeCost < dist[nextNode][usedSwitch]) {
                    dist[nextNode][usedSwitch] = cost + edgeCost;
                    pq.offer(new int[]{dist[nextNode][usedSwitch], nextNode, usedSwitch});
                }
            }
            
            // Use switch to reverse edges
            if (usedSwitch == 0) {
                for (int[] edge : edges) {
                    if (edge[1] == node) {
                        int nextNode = edge[0], edgeCost = edge[2];
                        if (cost + 2 * edgeCost < dist[nextNode][1]) {
                            dist[nextNode][1] = cost + 2 * edgeCost;
                            pq.offer(new int[]{dist[nextNode][1], nextNode, 1});
                        }
                    }
                }
            }
        }
        
        return -1;
    }
}

File: Medium/2026-01-27-3650-Minimum-Cost-Path-with-Edge-Reversals/solution.py
import heapq
from collections import defaultdict

class Solution:
    def minimumCost(self, n: int, edges: List[List[int]]) -> int:
        graph = defaultdict(list)
        for u, v, w in edges:
            graph[u].append((v, w))
            graph[v].append((u, 2 * w))  # Reversed edge with cost 2 * w
        
        min_cost = [float('inf')] * n
        min_cost[0] = 0
        pq = [(0, 0)]  # (cost, node)
        
        while pq:
            cost, node = heapq.heappop(pq)
            if cost > min_cost[node]:
                continue
            
            for neighbor, edge_cost in graph[node]:
                new_cost = cost + edge_cost
                if new_cost < min_cost[neighbor]:
                    min_cost[neighbor] = new_cost
                    heapq.heappush(pq, (new_cost, neighbor))
        
        return min_cost[n - 1] if min_cost[n - 1] != float('inf') else -1